src/adapter/src/client.rs

```diff
@@ -1032,3 +1039,62 @@ impl RecordFirstRowStream {
         msg
     }
 }
+
+/// ~~SPOOKY ZONE~~
+///
+/// There is a small race that can occur with [`tokio`], a [`SessionClient`], and the `Coordinator`
+/// that requires this Drop guard.
+///
+/// 1. A [`SessionClient`] can execute a command, sending it's owned [`Session`] to the
+///    Coordinator.
+/// 2. While processing this request, the user closes their connection to Materialize.
+/// 3. The Coordinator (which is run concurrently on a different thread and in a different task)
+///    finishes processing our request and is able to successfully send a response over the
+///    provided [`oneshot::Sender`].
+/// 4. [`tokio`] realizes the connection is closed and Drops the task which is awaiting the result
+///    from the [`oneshot::Receiver`].
+///
+///    !!Problem!! The Coordinator successfully responded to the request, so it doesn't need to
+///                terminate the [`Session`], but at the same time Future that is responsible for
+///                listening to the Receiver and putting the [`Session`] back into the
+///                [`SessionClient`] doesn't get polled, so the [`SessionClient`] never gets
+///                updated. Now neither the [`SessionClient`] nor the Coordinator believe they need
+///                (or can) terminate the [`Session`] so it effectively gets "leaked".
+///
+///    !!Solution!! This guard holds a mutable reference to the [`SessionClient`] and the
+///                 [`oneshot::Receiver`] that will return our [`Session`]. This way if we're ever
+///                 dropped mid-processing we can check for this race and re-assign the [`Session`]
+///                 to the [`SessionClient`] so it can properly terminate the session as part of
+///                 its Drop impl.
+pub struct SessionTerminateGuard<'a, T> {
+    client: &'a mut SessionClient,
+    rx: oneshot::Receiver<Response<T>>,
+}
+
+impl<'a, T> Future for SessionTerminateGuard<'a, T> {
+    type Output = Result<Response<T>, tokio::sync::oneshot::error::RecvError>;
+
+    fn poll(
+        mut self: pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
```

Comment:
Can you `use std::task::{Context, Poll}` and `use std::pin::Pin` here?

Reply:
Done!